src/main/java/gift/order/service/KakaoMessageService.java

```diff
@@ -0,0 +1,108 @@
+import org.springframework.web.client.RestClient;
+
+@Service
+public class KakaoMessageService {
+    private final RestClient restClient;
+    private final String talkUri;
+    private final ObjectMapper objectMapper;
+    private final UserKakaoTokenRepository userKakaoTokenRepository;
+    private final KakaoLoginService kakaoLoginService;
+
+    public KakaoMessageService(@Value("${kakao.talk_uri}") String talkUri,
+                               ObjectMapper objectMapper,
+                               UserKakaoTokenRepository userKakaoTokenRepository, KakaoLoginService kakaoLoginService) {
+        this.talkUri = talkUri;
+        this.objectMapper = objectMapper;
+        this.userKakaoTokenRepository = userKakaoTokenRepository;
+        this.kakaoLoginService = kakaoLoginService;
+
+        SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
+        requestFactory.setConnectTimeout(3000);
+        requestFactory.setReadTimeout(3000);
+
+        this.restClient = RestClient.builder()
+                .requestFactory(requestFactory)
+                .build();
+    }
+
+    public void sendMessage(Order order){
+        Member member = order.getMember();
+
+        String accessToken = userKakaoTokenRepository.findById(member.getId())
+                .orElseThrow(() -> new IllegalArgumentException("카카오 엑세스 토큰 조회에 실패하였습니다."))
+                .getAccessToken();
+
+        if (!kakaoLoginService.checkTalkMessageAgree(accessToken)) {
+            throw new IllegalStateException("사용자가 카카오톡 메시지 수신에 동의하지 않았습니다.");
+        }
+
+        String text = createText(order);
+
+        MessageLinkDto messageLinkDto = new MessageLinkDto("http://spring-product-web.com",
+                "http://m.spring-product-web.com");
+
+        MessageTemplateDto messageTemplateDto = new MessageTemplateDto("text", text, messageLinkDto);
+
+        try{
+            String jsonTemplate = objectMapper.writeValueAsString(messageTemplateDto);
+
+            MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
+
+            body.add("template_object", jsonTemplate);
+
+            restClient.post()
+                    .uri(talkUri)
+                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + accessToken)
+                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+                    .body(body)
+                    .retrieve()
+                    .onStatus(HttpStatusCode::is4xxClientError, ((request, response) -> {
+                        throw new KakaoClientException("카카오에 잘못된 템플릿을 전달하였습니다.");
+                    }))
+                    .onStatus(HttpStatusCode::is5xxServerError, ((request, response) -> {
+                        throw new KakaoServerException("카카오 서버에 문제가 발생했습니다.");
+                    }))
+                    .toBodilessEntity();
+
+        }catch(JsonProcessingException e){
+            throw new IllegalArgumentException("카카오 메시지 생성에 실패하였습니다.");
+        }
```

Comment:
```suggestion
        try{
            String jsonTemplate = objectMapper.writeValueAsString(messageTemplateDto);
        }catch(JsonProcessingException e){
            throw new IllegalArgumentException("카카오 메시지 생성에 실패하였습니다.");
        }

            MultiValueMap<String, String> body = new LinkedMultiValueMap<>();

            body.add("template_object", jsonTemplate);

            restClient.post()
                    .uri(talkUri)
                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + accessToken)
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .body(body)
                    .retrieve()
                    .onStatus(HttpStatusCode::is4xxClientError, ((request, response) -> {
                        throw new KakaoClientException("카카오에 잘못된 템플릿을 전달하였습니다.");
                    }))
                    .onStatus(HttpStatusCode::is5xxServerError, ((request, response) -> {
                        throw new KakaoServerException("카카오 서버에 문제가 발생했습니다.");
                    }))
                    .toBodilessEntity();
```
혹시 이 예외는 jsonTemplate 만들때만 발생할까요? 그럼 해당 블록만 try로 감싸면 어떨까요?

Reply:
넵! 멘토님의 말씀대로 해당 예외는 objectMapper 를 통해 만들 때만 발생하기 때문에 try-catch 블록을 좁히는 게 더 좋은 코드라고 생각해 수정하였습니다.